improver/psychrometric_calculations/hail_size.py

```diff
@@ -0,0 +1,295 @@
+        )
+        ccl_level_temperature = temperature_at_268.copy(data=t_dry)
+        ccl_level_temperature.rename("potential temperature of 268K at ccl pressure")
+        return ccl_level_temperature
+
+    def get_hail_size(self, vertical, horizontal):
+
+        """Uses the lookup_table and the vertical and horizontal indexes calculated to extract and store values from nomogram """
+
+        lookup_table = self.nomogram_values()
+        shape = np.shape(vertical)
+
+        vertical = np.around(vertical.data / 5, decimals=0) * 5
+        horizontal = np.around(horizontal.data * 2, decimals=0)
+
+        vertical_flat = vertical.flatten(order="C")
+        horizontal_flat = horizontal.flatten(order="C")
+
+        hail_size_list = []
+
+        for vertical, horizontal in zip(vertical_flat, horizontal_flat):
+
+            if horizontal > len(lookup_table[vertical]):
+                horizontal = len(lookup_table[vertical])
+
+            hail_size_list.append(lookup_table[int(vertical)][int(horizontal - 1)])
+
+        hail_size = np.reshape(hail_size_list, shape, order="C")
+
+        return hail_size
+
+    def make_hail_cube(self, hail_size, ccl_temperature, ccl_pressure):
+
+        """makes a cube containing hail_data"""
+
+        attributes = {}
+        if self.model_id_attr:
+            attributes[self.model_id_attr] = self.ccl_temperature.attributes[
+                self.model_id_attr
+            ]
+
+        hail_size_cube = create_new_diagnostic_cube(
+            name="size_of_hail_stones",
+            units="mm",
+            template_cube=ccl_temperature,
+            data=hail_size,
+            mandatory_attributes=generate_mandatory_attributes(
+                [ccl_temperature, ccl_pressure]
+            ),
+            optional_attributes=attributes,
+        )
+
+        return hail_size_cube
+
+    def process(
+        self,
+        ccl_temperature,
+        ccl_pressure,
+        temperature_on_pressure,
+        humidity_mixing_ratio_on_pressure,
+    ) -> Cube:
+
+        atmosphere_temperature_at_CCL = self.temperature_of_atmosphere_at_CCL_pressure(
+            temperature_on_pressure, ccl_pressure, ccl_temperature
+        )
+        temperature_at_268, pressure_at_268 = self.pressure_of_atmosphere_at_268(
+            temperature_on_pressure
+        )
+        humidity_mixing_ratio_at_CCL_pressure = self.humidity_mixing_ratio_at_CCL(
+            humidity_mixing_ratio_on_pressure, pressure_at_268
+        )
+
+        B = temperature_at_268
+        c = self.Dry_adiabatic_drop_to_ccl(
+            ccl_pressure, temperature_at_268, pressure_at_268
+        )
+        b = self.potential_temperature_after_saturated_ascent_from_CCL(
+            atmosphere_temperature_at_CCL,
+            ccl_pressure,
+            pressure_at_268,
+            humidity_mixing_ratio_at_CCL_pressure,
+            ccl_temperature,
+        )
+
+        vertical = c - B
+        horizontal = b - B
```

Comment:
If you are going to use the single character names from the paper, we need to have very clear comments about what they are.